File: scripts/generate_translation_availability.py
import os
import json

TRANSLATIONS_DIR = "src/data/translations"
OUTPUT_FILE = "src/data/translationAvailability.json"

def main():
    availability = {} # surah_id (str) -> list of "lang-author"
    
    # Initialize for all 114 surahs to ensure no undefined keys
    for i in range(1, 115):
        availability[str(i)] = []

    if not os.path.exists(TRANSLATIONS_DIR):
        print(f"Directory {TRANSLATIONS_DIR} does not exist.")
        return

    # Iterate languages
    for lang in os.listdir(TRANSLATIONS_DIR):
        lang_path = os.path.join(TRANSLATIONS_DIR, lang)
        if not os.path.isdir(lang_path):
            continue
            
        # Iterate authors
        for author in os.listdir(lang_path):
            author_path = os.path.join(lang_path, author)
            if not os.path.isdir(author_path):
                continue
                
            # Iterate files
            for filename in os.listdir(author_path):
                if filename.endswith(".yaml"):
                    try:
                        # filename format: 001-Subject.yaml
                        surah_id_str = filename.split('-')[0]
                        surah_id = str(int(surah_id_str)) # remove leading zeros
                        
                        key = f"{lang}-{author}"
                        
                        if surah_id in availability:
                            if key not in availability[surah_id]:
                                availability[surah_id].append(key)
                        else:
                            # Should have been initialized but just in case
                            availability[surah_id] = [key]
                            
                    except ValueError:
                        print(f"Skipping invalid filename: {filename}")

    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(availability, f, indent=2)
    
    print(f"Generated {OUTPUT_FILE}")

if __name__ == "__main__":
    main()

File: scripts/generate_translation_availability.js
import fs from 'fs';
import path from 'path';

const TRANSLATIONS_DIR = path.resolve('src/data/translations');
const OUTPUT_FILE = path.resolve('src/data/translationAvailability.json');

const main = () => {
    const availability = {};

    // Initialize for all 114 surahs
    for (let i = 1; i <= 114; i++) {
        availability[i] = [];
    }

    // Scan directories
    if (fs.existsSync(TRANSLATIONS_DIR)) {
        const languages = fs.readdirSync(TRANSLATIONS_DIR).filter(f => fs.statSync(path.join(TRANSLATIONS_DIR, f)).isDirectory());

        for (const lang of languages) {
            const langDir = path.join(TRANSLATIONS_DIR, lang);
            const authors = fs.readdirSync(langDir).filter(f => fs.statSync(path.join(langDir, f)).isDirectory());

            for (const author of authors) {
                const authorDir = path.join(langDir, author);
                const files = fs.readdirSync(authorDir).filter(f => f.endsWith('.yaml'));

                for (const file of files) {
                    // Extract Surah ID from filename "001-Al-Fatiha.yaml" -> 1
                    const match = file.match(/^(\d+)-/);
                    if (match) {
                        const surahId = parseInt(match[1], 10);
                        const key = `${lang}-${author}`;

                        if (availability[surahId]) {
                            availability[surahId].push(key);
                        }
                    }
                }
            }
        }
    }

    // Sort entries
    for (const id in availability) {
        availability[id].sort();
    }

    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(availability, null, 2));
    console.log(`Generated ${OUTPUT_FILE}`);
};

main();
